### PR TITLE
[SPARK-38361][SQL] Add factory method `getConnection` into `JDBCDialect`.

### PR DESCRIPTION
**Pull request by beliefer**
### What changes were proposed in this pull request?
At present, the parameter of the factory method for obtaining JDBC connection is empty because the JDBC URL of some databases is fixed and unique.
However, for databases such as ClickHouse, connection is related to the shard node.
So I think the parameter form of `getConnection: Partition = > Connection` is more general.

This PR adds factory method `getConnection` into `JDBCDialect` according to https://github.com/apache/spark/pull/35696#issuecomment-1058060107.

### Why are the changes needed?
Make factory method `getConnection` more general.

### Does this PR introduce _any_ user-facing change?
'No'.
Just inner change.

### How was this patch tested?
Exists test.

### CODE REVIEW

**Comment by beliefer** [comment]
ping @huaxingao cc @cloud-fan

File: sql/core/src/main/scala/org/apache/spark/sql/jdbc/JdbcDialects.scala [hunk 1]
```diff
@@ -99,6 +99,16 @@ abstract class JdbcDialect extends Serializable with Logging{
...
   def getJDBCType(dt: DataType): Option[JdbcType] = None
 
+  /**
+   * Get the factory method for create JDBC connection.
+   * In general, creating a connection has nothing to do with JDBC partition.
+   * But sometimes it is needed, such as a database with multiple shard nodes.
+   * @param options JDBC options.
+   * @return The factory method for create JDBC connection.
+   */
+  def getConnection(options: JDBCOptions): JDBCPartition => Connection =
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/jdbc/JdbcDialects.scala:109]: This is a developer-facing API, do we really want to expose `JDBCPartition` to users?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCRDD.scala [hunk 2]
```diff
@@ -204,7 +204,7 @@ object JDBCRDD extends Logging {
  */
 private[jdbc] class JDBCRDD(
     sc: SparkContext,
-    getConnection: () => Connection,
+    getConnection: JDBCPartition => Connection,
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCRDD.scala:207]: What partition does spark pass to this function if spark wants to create a connection at the driver side?

File: sql/core/src/main/scala/org/apache/spark/sql/jdbc/JdbcDialects.scala (hunk 1 above)
Comment (beliefer) [reply sql/core/src/main/scala/org/apache/spark/sql/jdbc/JdbcDialects.scala:109]: Yes. Users could use the JDBCPartition to get connection with different partition.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCRDD.scala (hunk 2 above)
Comment (beliefer) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCRDD.scala:207]: Users could construct JDBCPartition by self way. Such as: set the partition index with shard node list.

File: sql/core/src/main/scala/org/apache/spark/sql/jdbc/JdbcDialects.scala [hunk 3]
```diff
@@ -99,6 +99,16 @@ abstract class JdbcDialect extends Serializable with Logging{
...
+    JDBCPartition => JdbcUtils.createConnectionFactory(options)()
```
[the remark this answers is not in the document]
Comment (beliefer) [reply sql/core/src/main/scala/org/apache/spark/sql/jdbc/JdbcDialects.scala:110]: `JdbcUtils.createConnectionFactory` are used in other places. Should we remove `createConnectionFactory` from `JdbcUtils` or copy the code from `createConnectionFactory` to here?

File: sql/core/src/main/scala/org/apache/spark/sql/jdbc/JdbcDialects.scala (hunk 1 above)
Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/jdbc/JdbcDialects.scala:109]: what information do we need exactly? If you need partition id, isn't `Int` good enough here?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCRDD.scala (hunk 2 above)
Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCRDD.scala:207]: This did not answer the question.

File: sql/core/src/main/scala/org/apache/spark/sql/jdbc/JdbcDialects.scala (hunk 3 above)
[the remark this answers is not in the document]
Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/jdbc/JdbcDialects.scala:110]: > JdbcUtils.createConnectionFactory are used in other places

This is the problem. It's quite confusing that in some places Spark decides how to create the connection, and in some places JDBC dialect decides it.

File: sql/core/src/main/scala/org/apache/spark/sql/jdbc/JdbcDialects.scala (hunk 1 above)
Comment (beliefer) [reply sql/core/src/main/scala/org/apache/spark/sql/jdbc/JdbcDialects.scala:109]: Yes. we need partition id.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCRDD.scala (hunk 2 above)
Comment (beliefer) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCRDD.scala:207]: Yeah. In general, Spark create a connection at Driver side no need Partition or not know which Partition is.

File: sql/core/src/main/scala/org/apache/spark/sql/jdbc/JdbcDialects.scala (hunk 3 above)
[the remark this answers is not in the document]
Comment (beliefer) [reply sql/core/src/main/scala/org/apache/spark/sql/jdbc/JdbcDialects.scala:110]: OK. Let's use Option[JDBCPartition] as parameter.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala [hunk 4]
```diff
@@ -54,23 +53,17 @@ import org.apache.spark.util.NextIterator
...
  */
 object JdbcUtils extends Logging with SQLConfHelper {
+
   /**
    * Returns a factory for creating connections to the given JDBC URL.
    *
    * @param options - JDBC options that contains url, table and other information.
    * @throws IllegalArgumentException if the driver could not open a JDBC connection.
    */
   def createConnectionFactory(options: JDBCOptions): () => Connection = {
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala:63]: how do we guarantee this is only called at the driver side?

File: sql/core/src/main/scala/org/apache/spark/sql/jdbc/JdbcDialects.scala [hunk 5]
```diff
@@ -99,6 +100,27 @@ abstract class JdbcDialect extends Serializable with Logging{
...
 
+  /**
+   * Returns a factory for creating connections to the given JDBC URL.
+   * In general, creating a connection has nothing to do with JDBC partition id.
+   * But sometimes it is needed, such as a database with multiple shard nodes.
+   * @param options - JDBC options that contains url, table and other information.
+   * @return The factory method for creating JDBC connections.
+   * @throws IllegalArgumentException if the driver could not open a JDBC connection.
+   */
+  def getConnection(options: JDBCOptions): Int => Connection = {
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/jdbc/JdbcDialects.scala:111]: createConnectionFactory

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala (hunk 4 above)
Comment (beliefer) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala:63]: Could we dependent `SparkConf` here?

File: sql/core/src/main/scala/org/apache/spark/sql/jdbc/JdbcDialects.scala [hunk 6]
```diff
@@ -99,6 +100,27 @@ abstract class JdbcDialect extends Serializable with Logging{
    */
   def getJDBCType(dt: DataType): Option[JdbcType] = None
 
+  /**
+   * Returns a factory for creating connections to the given JDBC URL.
+   * In general, creating a connection has nothing to do with JDBC partition id.
+   * But sometimes it is needed, such as a database with multiple shard nodes.
+   * @param options - JDBC options that contains url, table and other information.
+   * @return The factory method for creating JDBC connections.
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/jdbc/JdbcDialects.scala:108]:
```suggestion
   * @return The factory method for creating JDBC connections with the RDD partition ID. -1 means
           the connection is being created at the driver side.
```

**Review by cloud-fan — APPROVED** [review]

File: sql/core/src/main/scala/org/apache/spark/sql/jdbc/JdbcDialects.scala [hunk 7]
```diff
@@ -99,6 +100,28 @@ abstract class JdbcDialect extends Serializable with Logging{
...
+  /**
+   * Returns a factory for creating connections to the given JDBC URL.
+   * In general, creating a connection has nothing to do with JDBC partition id.
+   * But sometimes it is needed, such as a database with multiple shard nodes.
+   * @param options - JDBC options that contains url, table and other information.
+   * @return The factory method for creating JDBC connections with the RDD partition ID. -1 means
+             the connection is being created at the driver side.
+   * @throws IllegalArgumentException if the driver could not open a JDBC connection.
+   */
+  def createConnectionFactory(options: JDBCOptions): Int => Connection = {
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/jdbc/JdbcDialects.scala:112]: add `@Since("3.3.0")`

**Review by cloud-fan — APPROVED** [review]

File: sql/core/src/main/scala/org/apache/spark/sql/jdbc/JdbcDialects.scala (hunk 7 above)
Comment (beliefer) [reply sql/core/src/main/scala/org/apache/spark/sql/jdbc/JdbcDialects.scala:112]: OK

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Comment by beliefer** [comment]
@cloud-fan Thank you a lot!

**Comment by bb-ivo** [comment]
@beliefer 
@chenzhx 
@cloud-fan 

Is there a migration guide available for this change?
We are using `JdbcUtils.createConnectionFactory(..)` in our own company code, which is moved to another class by this PR and we are not 100% sure on how to migrate this.

**Comment by beliefer** [comment]
@bb-ivo 
```
         val dialect = JdbcDialects.get(options.url)
         val conn = dialect.createConnectionFactory(options)(-1)
```

**Comment by bb-ivo** [comment]
Thanks 👍

**Pull request closed without merging**